src/theory/bags/bags_rewriter.cpp

```diff
@@ -0,0 +1,427 @@
+    {
+      // (difference_subtract A (union_disjoint A B)) = emptybag
+      // (difference_subtract A (union_disjoint B A)) = emptybag
+      // (difference_subtract A (union_max A B)) = emptybag
+      // (difference_subtract A (union_max B A)) = emptybag
+      NodeManager* nm = NodeManager::currentNM();
+      Node emptyBag = nm->mkConst(EmptyBag(n.getType()));
+      return RewriteResponse(REWRITE_DONE, emptyBag);
+    }
+  }
+
+  if (n[0].getKind() == INTERSECTION_MIN)
+  {
+    if (n[1] == n[0][0] || n[1] == n[0][1])
+    {
+      // (difference_subtract (intersection_min A B) A) = emptybag
+      // (difference_subtract (intersection_min B A) A) = emptybag
+      NodeManager* nm = NodeManager::currentNM();
+      Node emptyBag = nm->mkConst(EmptyBag(n.getType()));
+      return RewriteResponse(REWRITE_DONE, emptyBag);
+    }
+  }
+
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteDifferenceRemove(const TNode& n) const
+{
+  Assert(n.getKind() == DIFFERENCE_REMOVE);
+
+  if (n[0].getKind() == EMPTYBAG || n[1].getKind() == EMPTYBAG)
+  {
+    // (difference_remove A emptybag) = A
+    // (difference_remove emptybag B) = emptybag
+    return RewriteResponse(REWRITE_DONE, n[0]);
+  }
+
+  if (n[0] == n[1])
+  {
+    // (difference_remove A A) = emptybag
+    NodeManager* nm = NodeManager::currentNM();
+    Node emptyBag = nm->mkConst(EmptyBag(n.getType()));
+    return RewriteResponse(REWRITE_DONE, emptyBag);
+  }
+
+  if (n[1].getKind() == UNION_DISJOINT || n[1].getKind() == UNION_MAX)
+  {
+    if (n[0] == n[1][0] || n[0] == n[1][1])
+    {
+      // (difference_remove A (union_disjoint A B)) = emptybag
+      // (difference_remove A (union_disjoint B A)) = emptybag
+      // (difference_remove A (union_max A B)) = emptybag
+      // (difference_remove A (union_max B A)) = emptybag
+      NodeManager* nm = NodeManager::currentNM();
+      Node emptyBag = nm->mkConst(EmptyBag(n.getType()));
+      return RewriteResponse(REWRITE_DONE, emptyBag);
+    }
+  }
+
+  if (n[0].getKind() == INTERSECTION_MIN)
+  {
+    if (n[1] == n[0][0] || n[1] == n[0][1])
+    {
+      // (difference_remove (intersection_min A B) A) = emptybag
+      // (difference_remove (intersection_min B A) A) = emptybag
+      NodeManager* nm = NodeManager::currentNM();
+      Node emptyBag = nm->mkConst(EmptyBag(n.getType()));
+      return RewriteResponse(REWRITE_DONE, emptyBag);
+    }
+  }
+
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteChoose(const TNode& n) const
+{
+  Assert(n.getKind() == BAG_CHOOSE);
+  if (n[0].getKind() == MK_BAG && n[0][1].isConst())
+  {
+    // (bag.choose (mkBag x c)) = x where c is a constant > 0
+    return RewriteResponse(REWRITE_DONE, n[0][0]);
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteCard(const TNode& n) const
+{
+  Assert(n.getKind() == BAG_CARD);
+  if (n[0].getKind() == EMPTYBAG)
```

Comment:
Ditto, this should be generalized to when the argument is constant.

Reply:
I agree this should be moved to `NormalForm::getNormalForm` function. For now it is removed from here